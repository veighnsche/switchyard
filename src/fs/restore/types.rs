//! Types for restore subsystem

#[derive(Clone, Copy, Debug, Default)]
pub struct RestoreStats {
    pub fsync_ms: u64,
}

/// Which snapshot pair to select for restore
#[derive(Clone, Copy, Debug)]
pub enum SnapshotSel {
    Latest,
    Previous,
}

/// Options for restore behavior
#[derive(Clone, Debug)]
pub struct RestoreOptions {
    pub dry_run: bool,
    pub force_best_effort: bool,
    pub backup_tag: String,
}

/// Prior state kind as encoded in sidecar
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum PriorKind {
    File,
    Symlink,
    None,
    Other(String),
}

impl PriorKind {
    pub fn from_str(s: &str) -> Self {
        match s {
            "file" => PriorKind::File,
            "symlink" => PriorKind::Symlink,
            "none" => PriorKind::None,
            other => PriorKind::Other(other.to_string()),
        }
    }
}
