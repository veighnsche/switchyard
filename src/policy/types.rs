use std::path::PathBuf;

#[derive(Clone, Copy, Debug)]
pub enum RiskLevel {
    Stop,
    Warn,
    Allow,
}

#[derive(Clone, Copy, Debug)]
pub enum ExdevPolicy {
    Fail,
    DegradedFallback,
}

impl Default for ExdevPolicy {
    fn default() -> Self { ExdevPolicy::Fail }
}

#[derive(Clone, Copy, Debug)]
pub enum LockingPolicy {
    Required,
    Optional,
}

#[derive(Clone, Copy, Debug)]
pub enum PreservationPolicy {
    Off,
    RequireBasic,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum SourceTrustPolicy {
    RequireTrusted,
    WarnOnUntrusted,
    AllowUntrusted,
}

#[derive(Clone, Copy, Debug)]
pub enum SmokePolicy {
    Off,
    Require { auto_rollback: bool },
}

#[derive(Clone, Debug, Default)]
pub struct Scope {
    pub allow_roots: Vec<PathBuf>,
    pub forbid_paths: Vec<PathBuf>,
}

#[derive(Clone, Debug)]
pub struct Rescue {
    pub require: bool,
    pub exec_check: bool,
    pub min_count: usize,
}

impl Default for Rescue {
    fn default() -> Self {
        Self { require: false, exec_check: false, min_count: 0 }
    }
}

#[derive(Clone, Debug)]
pub struct Risks {
    pub suid_sgid: RiskLevel,
    pub hardlinks: RiskLevel,
    pub source_trust: SourceTrustPolicy,
    pub ownership_strict: bool,
}

impl Default for Risks {
    fn default() -> Self {
        Self { suid_sgid: RiskLevel::Stop, hardlinks: RiskLevel::Stop, source_trust: SourceTrustPolicy::RequireTrusted, ownership_strict: false }
    }
}

#[derive(Clone, Debug)]
pub struct Durability {
    pub backup_durability: bool,
    pub sidecar_integrity: bool,
    pub preservation: PreservationPolicy,
}

impl Default for Durability {
    fn default() -> Self {
        Self { backup_durability: true, sidecar_integrity: true, preservation: PreservationPolicy::Off }
    }
}

#[derive(Clone, Debug, Default)]
pub struct ApplyFlow {
    pub exdev: ExdevPolicy,
    pub override_preflight: bool,
    pub best_effort_restore: bool,
    pub extra_mount_checks: Vec<PathBuf>,
    pub capture_restore_snapshot: bool,
}

#[derive(Clone, Debug)]
pub struct Governance {
    pub locking: LockingPolicy,
    pub smoke: SmokePolicy,
    pub allow_unlocked_commit: bool,
}

impl Default for Governance {
    fn default() -> Self {
        Self { locking: LockingPolicy::Optional, smoke: SmokePolicy::Off, allow_unlocked_commit: true }
    }
}

#[derive(Clone, Debug, Default)]
pub struct Backup {
    pub tag: String,
}
