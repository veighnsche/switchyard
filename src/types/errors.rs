use thiserror::Error;

#[derive(Debug, Error)]
pub enum ErrorKind {
    #[error("invalid path")]
    InvalidPath,
    #[error("io error")]
    Io,
    #[error("policy violation")]
    Policy,
}

#[derive(Debug, Error)]
#[error("{kind:?}: {msg}")]
pub struct Error {
    pub kind: ErrorKind,
    pub msg: String,
}

pub type Result<T> = std::result::Result<T, Error>;
