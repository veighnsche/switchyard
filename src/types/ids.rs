//! Deterministic UUIDv5 identifiers for plans and actions.
//!
//! The UUID namespace is derived from a stable tag (`NS_TAG`) so that
//! `plan_id` and `action_id` are reproducible across runs for the same
//! serialized action sequence.
use std::fmt::Write;
use uuid::Uuid;

use super::{
    plan::{Action, Plan},
    safepath::SafePath,
};
// UUIDv5 namespace tag for deterministic plan/action IDs.
// See SPEC Reproducible v1.1 (Determinism) for guidance.
use crate::constants::NS_TAG;

/// Internal: return the UUID namespace used for deterministic IDs.
fn namespace() -> Uuid {
    Uuid::new_v5(&Uuid::NAMESPACE_URL, NS_TAG.as_bytes())
}

/// Return the relative representation of a `SafePath` to keep IDs independent of roots.
fn sp_rel(p: &SafePath) -> String {
    // Use the relative portion only for determinism across roots
    p.rel().to_string_lossy().to_string()
}

/// Serialize an action into a stable, human-readable string used for UUIDv5 input.
fn serialize_action(a: &Action) -> String {
    match a {
        Action::EnsureSymlink { source, target } => {
            format!("E:{}->{}", sp_rel(source), sp_rel(target))
        }
        Action::RestoreFromBackup { target } => {
            format!("R:{}", sp_rel(target))
        }
    }
}

/// Compute a deterministic UUIDv5 for a plan by serializing actions in order.
///
/// Two plans with identical action sequences (including ordering) will have the
/// same `plan_id`, independent of the root directories used by `SafePath`.
#[must_use]
pub fn plan_id(plan: &Plan) -> Uuid {
    let ns = namespace();
    // Deterministic serialization in action order
    let mut s = String::new();
    for a in &plan.actions {
        s.push_str(&serialize_action(a));
        s.push('\n');
    }
    Uuid::new_v5(&ns, s.as_bytes())
}

/// Compute a deterministic UUIDv5 for an action as a function of the plan ID and
/// the action's serialized form, including the stable position index.
#[must_use]
pub fn action_id(plan_id: &Uuid, action: &Action, idx: usize) -> Uuid {
    let mut s = serialize_action(action);
    let _ = write!(s, "#{idx}"); // Ignore formatting errors as they're unlikely in this context
    Uuid::new_v5(plan_id, s.as_bytes())
}
