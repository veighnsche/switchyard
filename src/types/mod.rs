pub mod errors;
pub mod ids;
pub mod mount;
pub mod ownership;
pub mod plan;
pub mod preflight;
pub mod report;
pub mod rescue;
pub mod safepath;

pub use errors::*;
pub use ids::*;
pub use mount::*;
pub use ownership::*;
pub use plan::*;
pub use preflight::*;
pub use report::*;
pub use rescue::*;
pub use safepath::*;
