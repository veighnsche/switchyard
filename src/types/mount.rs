//! Data-only mount types used across the crate.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MountFlags {
    pub read_only: bool,
    pub no_exec: bool,
}

#[derive(Debug, thiserror::Error, Clone)]
pub enum MountError {
    #[error("unknown or ambiguous mount state")]
    Unknown,
}
