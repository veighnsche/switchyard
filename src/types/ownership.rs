//! Data-only type for ownership information of a filesystem path.
//! Centralized under `crate::types` for cross-layer reuse.

#[derive(Clone, Debug)]
pub struct OwnershipInfo {
    pub uid: u32,
    pub gid: u32,
    pub pkg: String,
}
