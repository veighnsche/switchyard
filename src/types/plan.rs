use super::safepath::SafePath;

#[derive(Debug, Copy, Clone, Default)]
pub enum ApplyMode {
    #[default]
    DryRun,
    Commit,
}

#[derive(Clone, Debug)]
pub struct LinkRequest {
    pub source: SafePath,
    pub target: SafePath,
}

#[derive(Clone, Debug)]
pub struct RestoreRequest {
    pub target: SafePath,
}

#[derive(Clone, Debug, Default)]
pub struct PlanInput {
    pub link: Vec<LinkRequest>,
    pub restore: Vec<RestoreRequest>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Action {
    EnsureSymlink { source: SafePath, target: SafePath },
    RestoreFromBackup { target: SafePath },
}

#[derive(Clone, Debug, Default)]
pub struct Plan {
    pub actions: Vec<Action>,
}
