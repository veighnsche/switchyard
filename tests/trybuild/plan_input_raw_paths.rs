// compile-fail: constructing PlanInput with raw PathBuf instead of SafePath
use switchyard::types::{PlanInput, LinkRequest};

fn main() {
    let _input = PlanInput {
        link: vec![LinkRequest {
            // These should be switchyard::types::SafePath, not PathBuf
            source: std::path::PathBuf::from("/a"),
            target: std::path::PathBuf::from("/b"),
        }],
        restore: vec![],
    };
}
