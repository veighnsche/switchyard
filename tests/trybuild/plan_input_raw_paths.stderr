error[E0308]: mismatched types
 --> tests/trybuild/plan_input_raw_paths.rs:8:21
  |
8 |             source: std::path::PathBuf::from("/a"),
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `SafePath`, found `PathBuf`

error[E0308]: mismatched types
 --> tests/trybuild/plan_input_raw_paths.rs:9:21
  |
9 |             target: std::path::PathBuf::from("/b"),
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `SafePath`, found `PathBuf`
